Implement SVC

Renkooo opened the pull request:
This pull request addresses issue #14  : "Implement the SVC algorithm (SVM for classification task)".

**Implements features: **

**SVM Classification:** Added SVM Classification support in `svc.py`.
**Tests:** Updated `test_svc.py` with tests for SVM Classification functionality.
**Examples:** Added a usage example in` svc.py` for SVM Classification.

Looking forward to any feedback or suggestions.

rafaelgreca commented on examples/svc.py:
```diff
@@ -0,0 +1,70 @@
+from scratchml.models.svc import SVC
+from sklearn.datasets import make_classification
+from sklearn.model_selection import train_test_split, cross_val_score
```
We already have a function for that, please change this line to:

```python
from scratchml.utils import train_test_split
```

rafaelgreca commented on examples/svc.py:
```diff
@@ -0,0 +1,70 @@
+
+    # Split the dataset into training and test sets
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.2, random_state=42
```
Please, delete the ```random_state=42```, since we don't have this parameter in our implementation.

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+
+    _valid_metrics = ["accuracy", "precision", "recall", "f1_score", "confusion_matrix"]
+
+    def __init__(
```
Please, add typing for parameters and outputs for all functions. Here is an example of how to do it using that function:

```python3
def __init__(
   self,
   C: float = 0.35,
   alpha: float = 0.01,
   kernel: str = "linear",
   degree: int = 3,
   max_iter: int = 1500,
   tol: float = 1e-4,
   learning_rate: float = 1e-5,
   decay: float = 0.999,
   batch_size: int = 64,
   early_stopping: bool = True,
   adaptive_lr: bool = False,
) -> None:
```

Remember, for some types (like List, Tuple, and others), you need to import the built-in 'Typing' library. You can do that with the following code:

```python3
from typing import List, Tuple, Dict
```

rafaelgreca replied:
Please, add a docstring as well. Like this:

```python3
def init(...) -> None:
    """
    Creates a Support Vector Classifier (SVC) instance.

    Args:
       C (float): the parameter's description goes here. Defaults to 0.35.
       alpha (float): the parameter's description goes here. Defaults to 0.01.
       kernel (str): the parameter's description goes here. Put the valid parameters as well. Defaults to "linear".
       degree (int): the parameter's description goes here. Defaults to 3.
       max_iter (int): the parameter's description goes here. Defaults to 1500.
       tol (float): the parameter's description goes here. Defaults to 1e-4.
       learning_rate (float): the parameter's description goes here. Defaults to 1e-5.
       decay (float): the parameter's description goes here. Defaults to 0.999.
       batch_size (int): the parameter's description goes here. Defaults to 64.
       early_stopping (bool): the parameter's description goes here. Defaults to True.
       adaptive_lr (bool): the parameter's description goes here. Defaults to False.
    """
```

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+                -gamma * np.square(np.linalg.norm(X1[:, np.newaxis] - X2, axis=2))
+            )
+
+    def _validate_parameters(self):
```
Please, add typing here too. Like this:

```python3
def _validate_parameters(self) -> None:
```

rafaelgreca replied:
Please, add a docstring as well. Like this:

```python3
def _validate_parameters(self) -> None:
   """
   Auxiliary function used to validate the values of the parameters
   passed during the initialization.
   """
```

rafaelgreca replied:
Please, add a validation for other parameters as well (such as kernel, alpha, max_iter, degree, tol, and others). You can verify the parameter value and its type. Check out other implementations to get an idea of how to do that (e.g., for the Random Forest Classifier implementation, check line 241 in file random_forest.py).

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+        if self.C <= 0:
+            raise ValueError("Regularization parameter C must be positive.")
+
+    def _check_is_fitted(self):
```
Please, add typing here too. Like this:

```python3
def _check_is_fitted(self) -> None:
```

rafaelgreca replied:
Please, add a docstring as well. Like this:

```python3
def _check_is_fitted(self) -> None:
    """
    Check whether the SVC instance is fitted.
    """
```

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+        if self.weights is None:
+            raise ValueError("Model must be trained before prediction.")
+
+    def fit(self, X, y):
```
Please, add typing here too. Like this:

```python3
def fit(self, X: np.ndarray, y: np.ndarray) -> None:
```

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+        # print(f"Final weight norm: {np.linalg.norm(self.weights)}")
+        # print(f"Final bias: {self.bias}")
+
+    def predict(self, X):
```
Please, add typing here too. Like this:

```python3
def predict(self, X: np.ndarray) -> np.ndarray:
```

rafaelgreca replied:
Please, add a docstring as well. Like this:

```python3
def predict(self, X: np.ndarray) -> np.ndarray:
    """
    Uses the trained model to predict the classes of a given
    data points (also called features).

    Args:
       X (np.ndarray): the features.

    Returns:
       np.ndarray: the predicted classes.
    """
```

You can get the typings and the docstring of that function from other files, such as the Random Forest Classifier code (line 364 in file random_forest.py).

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+            "confusion_matrix": confusion_matrix(y, predictions),
+        }
+
+    def score(self, X, y, metric="accuracy", labels_cm=None, normalize_cm=False):
```
Please, add typing here too. Like this:

```python3
def score(
   self,
   X: np.ndarray,
   y: np.ndarray,
   metric: str = "accuracy",
   labels_cm: List = None,
   normalize_cm: bool = False,
) -> Union[np.float64, np.ndarray]:
```

You can get the typings and the docstring of that function from other files, such as the Random Forest Classifier code (line 392 in file random_forest.py).

rafaelgreca replied:
Example of the docstring:

```python3
def score(
    self,
    X: np.ndarray,
    y: np.ndarray,
    metric: str = "accuracy",
    labels_cm: List = None,
    normalize_cm: bool = False,
) -> Union[np.float64, np.ndarray]:
    """
    Uses the trained model to predict the classes of a given
    data points (also called features).

    Args:
       X (np.ndarray): the features array.
       y (np.ndarray): the labels array.
       metric (string): which metric should be used to assess
           the model's performance. Defaults to Accuracy.
       labels_cm (str, optional): which labels should be used to calculate
           the confusion matrix. If other metric is selected, then this
           parameter will be ignore. Defaults to None.
       normalize_cm (bool, optional): whether the confusion matrix should be
           normalized ('all', 'pred', 'true') or not. If other metric is selected,
           then this parameter will be ignore. Defaults to False.

    Returns:
        (np.float32, np.ndarray): the score achieved by the model
            or its confusion matrix.
    """
```

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+        else:
+            return predictions
+
+    def evaluate(self, X, y):
```
Please, delete this function, as you already have the `score` function that does the same thing and there's no function `evaluate` in the Scikit-learn's SVC implementation.

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+        early_stopping=True,
+        adaptive_lr=False,
+    ):
+        if C <= 0:
```
Please, delete this if condition. You already do this verification in the `_validate_parameters` function.

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+        self.scaler = StandardScaler()
+        self._validate_parameters()
+
+    def _apply_kernel(self, X1, X2):
```
Please, add a docstring as well. Like this.

```python3
def _apply_kernel(self, X1: np.ndarray, X2: np.ndarray) -> np.ndarray:
    """
    Applies the selected kernel function.

    Args:
       X1 (np.ndarray): the parameter's description goes here.
       X2 (np.ndarray): the parameter's description goes here.

    Returns:
      np.ndarray: the output's description goes here.
    """
```

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+        if len(self.classes_) == 2:
+            y_ = np.where(y == self.classes_[0], -1, 1)
+        else:
+            y_ = y
```
Please, change this line to:

```python3
y_ = y.copy()
```

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+            y_ = y
+
+        X = self.scaler.fit_transform(X)
+        n_samples, n_features = X.shape
```
Please, change this line to (cause you're not using the `n_samples` variable):

```python3
_, n_features = X.shape
```

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+                if (
+                    no_improvement_count > 10
+                ):  # Terminate if no improvement for 10 iterations
+                    # print(f"Converged after {iteration} iterations.")
```
Please, delete this line.

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+            else:
+                no_improvement_count = 0
+
+        #     if iteration % 100 == 0:
```
I liked the idea of printing the updates every X epoch, what about creating a verbose parameter to control that? You can get an idea of how to implement that by looking at the Logistic Regression implementation (line 202 in file logistic_regression.py).

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+        #     if iteration % 100 == 0:
+        #         print(f"Iteration {iteration}: Avg norm of batch updates = {avg_update_norm}")
+
+        # print(f"Final weight norm: {np.linalg.norm(self.weights)}")
```
Please, delete this line.

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+        #         print(f"Iteration {iteration}: Avg norm of batch updates = {avg_update_norm}")
+
+        # print(f"Final weight norm: {np.linalg.norm(self.weights)}")
+        # print(f"Final bias: {self.bias}")
```
Please, delete this line.

rafaelgreca commented on scratchml/models/svc.py:
```diff
@@ -0,0 +1,202 @@
+        self.weights = np.zeros(n_features)
+
+        no_improvement_count = 0
+        for iteration in range(self.max_iter):
```
I think it would be a good idea to put some comments throughout the `fit` function on some steps that you think it's important to understand the logic behind the code and the main idea of how the algorithm works.

rafaelgreca commented on tests/models/test_svc.py:
```diff
@@ -0,0 +1,121 @@
+from scratchml.scalers import StandardScaler
+from ..utils import generate_classification_dataset, repeat
+import unittest
+import numpy as np
```
Please, delete this line, as you're not using numpy here.

rafaelgreca commented on tests/models/test_svc.py:
```diff
@@ -0,0 +1,121 @@
+        """
+        # Use scaled data for both models
+        X, y = generate_classification_dataset(
+            n_samples=2000, n_features=4, n_classes=2
```
Please, increase the `n_classes` value.

rafaelgreca commented on tests/models/test_svc.py:
```diff
@@ -0,0 +1,121 @@
+        atol = math.floor(y.shape[0] * 0.1)
+        assert_allclose(sklearn_pred, custom_pred, atol=atol)
+        assert abs(sklearn_score - custom_score) / abs(sklearn_score) < 0.05
+
```
Please, add the `assert_equal(sklearn_svc.classes_, custom_svc.classes_)` validation to this function as well.

rafaelgreca commented on tests/models/test_svc.py:
```diff
@@ -0,0 +1,121 @@
+        assert abs(sklearn_score - custom_score) / abs(sklearn_score) < 0.05
+
+    @repeat(3)
+    def test_rbf_kernel(self):
```
It would be nice to add a few more validations, such as asserting the output type and shape of both implementations are equal, other parameters value, and so on. Check out other implementation tests to see if you get any idea to increment this.

rafaelgreca commented on tests/models/test_svc.py:
```diff
@@ -0,0 +1,121 @@
+        assert abs(sklearn_score - custom_score) / abs(sklearn_score) < 0.05
+
+    @repeat(3)
+    def test_multi_class_classification(self):
```
It would be nice to add a few more validations, such as asserting the output type and shape of both implementations are equal, other parameters value, and so on. Check out other implementation tests to see if you get any idea to increment this.

rafaelgreca commented on tests/models/test_svc.py:
```diff
@@ -0,0 +1,121 @@
+    """
+
+    @repeat(3)
+    def test_binary_classification(self):
```
It would be nice to add a few more validations, such as asserting the output type and shape of both implementations are equal, other parameters value, and so on. Check out other implementation tests to see if you get any idea to increment this.

rafaelgreca reviewed (changes requested):
Hi, @Renkooo! First of all, thank you so much for your contribution.

I added a few comments; feel free to reach out to me if you have any questions or suggestions. In a few of them, I used some examples to illustrate a point; you don't need to do exactly as I exemplified, you can use your creativity and suggest other things.

Have you attempted to run the functions in `test_svc.py`? Was the verification successful? Here, they failed when comparing the results with the Scikit-Learn implementation. To be honest, I kind of anticipated that since I saw that they use a variety of sophisticated optimization techniques (like Nystroem transformer and quadratic optimization) to streamline their code. One possible solution is to make minor adjustments to the code implementation (e.g., using the same default values for the parameters as in Scikit-learn), then raise the difference value criterion (0.05) and check to see if the tests pass. If they don't, then we decide if it's worth deleting these validations for now.

Sorry, I forgot to add to the GitHub workflow, but can you please use the [Black](https://github.com/psf/black) code formatter (you only need to install it and then run black . --safe on the root folder)? Thanks so much for your contribution! Once you review my suggestions and commit again, I'll merge the code.

Oh, before I forget, can you please check/complete the "SVC" on the "Roadmap" section in the "README" file? Also, feel free to create a paragraph about yourself (just like I did) in the "AUTHORS" file inside the "docs" folder (you can put it on the "OTHER CONTRIBUTORS" section).

Again, thanks so much for your contribution to the project! Whenever you want to, feel free to contribute to other issues! :D

Renkooo commented:
Hello @rafaelgreca, 

Thank you for detailed code review and I have implemented all of your suggestions and improvements such as:
  - Used Black to format the code.
  - Updated SVM section in roadmap in README.
  - Added myself to the AUTHORS.
  - Updated code based on your feedback.

Thank you for your guidance and for helping me improve the code! Looking forward to your response.
